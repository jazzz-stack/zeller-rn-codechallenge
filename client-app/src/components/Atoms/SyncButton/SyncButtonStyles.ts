import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  syncButton: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#fff',
    paddingVertical: 12,
    paddingHorizontal: 16,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: '#3498DB',
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.1,
    shadowRadius: 4,
  },
  disabledButton: {
    backgroundColor: '#f5f5f5',
    borderColor: '#ddd',
    elevation: 0,
    shadowOpacity: 0,
  },
  syncButtonText: {
    color: '#3498DB',
    fontSize: 16,
    fontWeight: '600',
    marginLeft: 8,
  },
  disabledText: {
    color: '#999',
  },
  spinning: {
    // Note: For actual spinning animation, you'd need to use Animated API
    // This is just a placeholder for the spinning effect
  },
});
