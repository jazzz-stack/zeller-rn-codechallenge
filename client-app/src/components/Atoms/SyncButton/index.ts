export {default} from './SyncButton';
export type {SyncButtonProps} from './SyncButton';
