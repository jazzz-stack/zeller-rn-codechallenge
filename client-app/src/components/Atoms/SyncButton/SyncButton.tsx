import React from 'react';
import {TouchableOpacity, TouchableOpacityProps} from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import Text from '../Text';
import {styles} from './SyncButtonStyles';

export interface SyncButtonProps extends Omit<TouchableOpacityProps, 'style'> {
  title?: string;
  isLoading?: boolean;
  disabled?: boolean;
  iconSize?: number;
  iconColor?: string;
  containerStyle?: any;
  textStyle?: any;
}

const SyncButton: React.FC<SyncButtonProps> = ({
  title = 'Sync',
  isLoading = false,
  disabled = false,
  iconSize = 25,
  iconColor = '#3498DB',
  containerStyle,
  textStyle,
  ...touchableProps
}) => {
  const isDisabled = disabled || isLoading;

  return (
    <TouchableOpacity
      style={[
        styles.syncButton,
        isDisabled && styles.disabledButton,
        containerStyle,
      ]}
      disabled={isDisabled}
      {...touchableProps}
    >
      <Icon
        name={isLoading ? 'refresh' : 'sync'}
        size={iconSize}
        color={isDisabled ? '#999' : iconColor}
        style={isLoading ? styles.spinning : undefined}
      />
      <Text style={[styles.syncButtonText, isDisabled && styles.disabledText, textStyle]}>
        {isLoading ? 'Syncing...' : title}
      </Text>
    </TouchableOpacity>
  );
};

export default SyncButton;
