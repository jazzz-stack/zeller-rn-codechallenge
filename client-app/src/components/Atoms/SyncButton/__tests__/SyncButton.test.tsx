import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react-native';
import SyncButton from '../SyncButton';

// Mock react-native-vector-icons
jest.mock('react-native-vector-icons/MaterialIcons', () => {
  const React = require('react');
  const {Text} = require('react-native');
  
  return ({name, size, color, style, testID}: any) =>
    React.createElement(Text, {
      testID: testID || 'mock-icon',
      style: [style, {fontSize: size, color}],
    }, name);
});

describe('SyncButton', () => {
  const defaultProps = {
    onPress: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    it('renders with default props', () => {
      render(<SyncButton {...defaultProps} />);
      
      expect(screen.getByText('Sync')).toBeTruthy();
      expect(screen.getByText('sync')).toBeTruthy(); // Icon name
    });

    it('renders with custom title', () => {
      render(<SyncButton {...defaultProps} title="Retry" />);
      
      expect(screen.getByText('Retry')).toBeTruthy();
      expect(screen.queryByText('Sync')).toBeNull();
    });

    it('renders loading state correctly', () => {
      render(<SyncButton {...defaultProps} isLoading={true} />);
      
      expect(screen.getByText('Syncing...')).toBeTruthy();
      expect(screen.getByText('refresh')).toBeTruthy(); // Icon name for loading
    });

    it('renders disabled state correctly', () => {
      render(<SyncButton {...defaultProps} disabled={true} />);
      
      expect(screen.getByText('Sync')).toBeTruthy();
      expect(screen.getByText('sync')).toBeTruthy();
    });

    it('renders with custom icon properties', () => {
      render(<SyncButton {...defaultProps} iconSize={30} iconColor="#FF0000" />);
      
      const icon = screen.getByText('sync');
      expect(icon.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            fontSize: 30,
            color: '#FF0000',
          }),
        ])
      );
    });
  });

  describe('Interaction', () => {
    it('calls onPress when pressed and not disabled', () => {
      render(<SyncButton {...defaultProps} />);
      
      const button = screen.getByText('Sync').parent;
      fireEvent.press(button);
      
      expect(defaultProps.onPress).toHaveBeenCalledTimes(1);
    });

    it('does not call onPress when disabled', () => {
      render(<SyncButton {...defaultProps} disabled={true} />);
      
      const button = screen.getByText('Sync').parent;
      fireEvent.press(button);
      
      expect(defaultProps.onPress).not.toHaveBeenCalled();
    });

    it('does not call onPress when loading', () => {
      render(<SyncButton {...defaultProps} isLoading={true} />);
      
      const button = screen.getByText('Syncing...').parent;
      fireEvent.press(button);
      
      expect(defaultProps.onPress).not.toHaveBeenCalled();
    });

    it('passes through additional TouchableOpacity props', () => {
      const testID = 'custom-sync-button';
      render(<SyncButton {...defaultProps} testID={testID} />);
      
      expect(screen.getByTestId(testID)).toBeTruthy();
    });
  });

  describe('Styling', () => {
    it('applies custom container style', () => {
      const customStyle = {backgroundColor: 'red'};
      const testID = 'sync-button-container';
      render(<SyncButton {...defaultProps} containerStyle={customStyle} testID={testID} />);
      
      const button = screen.getByTestId(testID);
      expect(button.props.style).toEqual(
        expect.objectContaining(customStyle)
      );
    });

    it('applies custom text style', () => {
      const customTextStyle = {fontSize: 20};
      render(<SyncButton {...defaultProps} textStyle={customTextStyle} />);
      
      const text = screen.getByText('Sync');
      const flattenedStyle = JSON.stringify(text.props.style);
      expect(flattenedStyle).toContain(JSON.stringify(customTextStyle));
    });

    it('applies disabled styles when disabled', () => {
      render(<SyncButton {...defaultProps} disabled={true} />);
      
      const icon = screen.getByText('sync');
      expect(icon.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            color: '#999',
          }),
        ])
      );
    });

    it('applies disabled styles when loading', () => {
      render(<SyncButton {...defaultProps} isLoading={true} />);
      
      const icon = screen.getByText('refresh');
      expect(icon.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            color: '#999',
          }),
        ])
      );
    });
  });

  describe('Accessibility', () => {
    it('is accessible by default', () => {
      render(<SyncButton {...defaultProps} />);
      
      const button = screen.getByText('Sync').parent;
      expect(button.props.accessible).not.toBe(false);
    });

    it('passes through accessibility props', () => {
      const accessibilityLabel = 'Sync data button';
      render(<SyncButton {...defaultProps} accessibilityLabel={accessibilityLabel} />);
      
      const button = screen.getByLabelText(accessibilityLabel);
      expect(button).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined title gracefully', () => {
      render(<SyncButton {...defaultProps} title={undefined} />);
      
      expect(screen.getByText('Sync')).toBeTruthy(); // Should use default
    });

    it('handles empty string title', () => {
      render(<SyncButton {...defaultProps} title="" />);
      
      expect(screen.queryByText('Sync')).toBeNull();
      // Should render empty string, but icon should still be there
      expect(screen.getByText('sync')).toBeTruthy();
    });

    it('handles both disabled and loading states', () => {
      render(<SyncButton {...defaultProps} disabled={true} isLoading={true} />);
      
      expect(screen.getByText('Syncing...')).toBeTruthy();
      expect(screen.getByText('refresh')).toBeTruthy();
      
      const button = screen.getByText('Syncing...').parent;
      fireEvent.press(button);
      
      expect(defaultProps.onPress).not.toHaveBeenCalled();
    });

    it('handles zero icon size', () => {
      render(<SyncButton {...defaultProps} iconSize={0} />);
      
      const icon = screen.getByText('sync');
      expect(icon.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            fontSize: 0,
          }),
        ])
      );
    });
  });

  describe('Snapshots', () => {
    it('matches snapshot in default state', () => {
      const component = render(<SyncButton {...defaultProps} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot in loading state', () => {
      const component = render(<SyncButton {...defaultProps} isLoading={true} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot in disabled state', () => {
      const component = render(<SyncButton {...defaultProps} disabled={true} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    it('matches snapshot with custom props', () => {
      const component = render(
        <SyncButton 
          {...defaultProps} 
          title="Custom Title"
          iconSize={40}
          iconColor="#FF5733"
          containerStyle={{backgroundColor: '#000'}}
          textStyle={{color: '#FFF'}}
        />
      );
      expect(component.toJSON()).toMatchSnapshot();
    });
  });
});
