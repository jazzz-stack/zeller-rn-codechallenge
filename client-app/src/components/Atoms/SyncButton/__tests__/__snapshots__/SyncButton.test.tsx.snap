// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncButton Snapshots matches snapshot in default state 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#3498DB",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 2,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        undefined,
        {
          "color": "#3498DB",
          "fontSize": 25,
        },
      ]
    }
    testID="mock-icon"
  >
    sync
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#000000",
        },
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "normal",
        },
        [
          {
            "color": "#3498DB",
            "fontSize": 16,
            "fontWeight": "600",
            "marginLeft": 8,
          },
          false,
          undefined,
        ],
      ]
    }
  >
    Sync
  </Text>
</View>
`;

exports[`SyncButton Snapshots matches snapshot in disabled state 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f5f5f5",
      "borderColor": "#ddd",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        undefined,
        {
          "color": "#999",
          "fontSize": 25,
        },
      ]
    }
    testID="mock-icon"
  >
    sync
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#000000",
        },
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "normal",
        },
        [
          {
            "color": "#3498DB",
            "fontSize": 16,
            "fontWeight": "600",
            "marginLeft": 8,
          },
          {
            "color": "#999",
          },
          undefined,
        ],
      ]
    }
  >
    Sync
  </Text>
</View>
`;

exports[`SyncButton Snapshots matches snapshot in loading state 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f5f5f5",
      "borderColor": "#ddd",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 0,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        {},
        {
          "color": "#999",
          "fontSize": 25,
        },
      ]
    }
    testID="mock-icon"
  >
    refresh
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#000000",
        },
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "normal",
        },
        [
          {
            "color": "#3498DB",
            "fontSize": 16,
            "fontWeight": "600",
            "marginLeft": 8,
          },
          {
            "color": "#999",
          },
          undefined,
        ],
      ]
    }
  >
    Syncing...
  </Text>
</View>
`;

exports[`SyncButton Snapshots matches snapshot with custom props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#000",
      "borderColor": "#3498DB",
      "borderRadius": 8,
      "borderWidth": 1,
      "elevation": 2,
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 4,
    }
  }
>
  <Text
    style={
      [
        undefined,
        {
          "color": "#FF5733",
          "fontSize": 40,
        },
      ]
    }
    testID="mock-icon"
  >
    sync
  </Text>
  <Text
    style={
      [
        {
          "fontFamily": "System",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        {
          "color": "#000000",
        },
        {
          "fontSize": 16,
        },
        {
          "fontWeight": "normal",
        },
        [
          {
            "color": "#3498DB",
            "fontSize": 16,
            "fontWeight": "600",
            "marginLeft": 8,
          },
          false,
          {
            "color": "#FFF",
          },
        ],
      ]
    }
  >
    Custom Title
  </Text>
</View>
`;
